src/main/java/cz/smarteon/loxone/app/state/LoxoneState.java

```diff
@@ -0,0 +1,74 @@
+package cz.smarteon.loxone.app.state;
+
+import cz.smarteon.loxone.*;
+import cz.smarteon.loxone.app.model.Control;
+import cz.smarteon.loxone.app.model.LoxoneApp;
+import cz.smarteon.loxone.app.model.SwitchControl;
+import cz.smarteon.loxone.message.TextEvent;
+import cz.smarteon.loxone.message.ValueEvent;
+import lombok.extern.slf4j.Slf4j;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class LoxoneState implements LoxoneAppListener, LoxoneEventListener {
+
+    private static final Map<Class<? extends Control>,Class<? extends ControlState<? extends Control>>> SUPPORTED_CONTROLS_STATE_MAP;
+
+    private final LoxoneWebSocket webSocket;
+
+    private LoxoneApp loxoneApp;
+
+    private Map<LoxoneUuid, ControlState<?>> controlStates;
+
+    static {
+        SUPPORTED_CONTROLS_STATE_MAP = new HashMap<>();
+        SUPPORTED_CONTROLS_STATE_MAP.put(SwitchControl.class, SwitchControlState.class);
+    }
+
+    public LoxoneState(LoxoneWebSocket webSocket) {
+        this.webSocket = webSocket;
+        webSocket.registerListener(this);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T> T getStateForControl(Control control) {
+        return (T) controlStates.get(control.getUuid());
+    }
+
+    @Override
+    public void onLoxoneApp(@NotNull LoxoneApp loxoneApp) {
+        this.loxoneApp = loxoneApp;
+        initializeState();
+    }
+
+    @Override
+    public void onEvent(@NotNull ValueEvent event) {
+        controlStates.values().forEach(cs -> cs.accept(event));
+    }
+
+    @Override
+    public void onEvent(@NotNull TextEvent event) {
+        controlStates.values().forEach(cs -> cs.accept(event));
+    }
+
+    private void initializeState() {
+        controlStates = loxoneApp.getControls().entrySet().stream()
+                .filter(e -> SUPPORTED_CONTROLS_STATE_MAP.containsKey(e.getValue().getClass()))
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        e -> {
+                            try {
+                                return (ControlState<?>) SUPPORTED_CONTROLS_STATE_MAP.get(e.getValue().getClass()).getConstructors()[0].newInstance(webSocket, e.getValue());
```

Comment:
any chance to avoid reflection?

Reply:
Will look into this when migrating to a factory class.